### PR TITLE
move nexus deployment to statefulset

### PR DESCRIPTION
**Pull request by cristifalcas**
**What this PR does / why we need it**:
This PR moves the deployment and PersistentVolumeClaim to statefulset and volumeClaimTemplates

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
With the addition of nexus-backup, deployment is useless now. It will not manage to crate 2 disks in the same zone and because of this it will never create the pod.

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by rjkernick** [comment]
Please sign the CLA and increase the chart version to at least 2.0.0

**Comment by cristifalcas** [comment]
Hi. I signed the CLA, but I don't know how to verify it. It will appear automatically here?

**Comment by rjkernick** [comment]
CLA is showing up

**Comment by rjkernick** [comment]
@jeff-knurek Can you take a look at this? Have you had any issues with the 2 pvc configuration?

**Comment by jeff-knurek** [comment]
hmm, no I haven't had any problem with disks not being created, but I've only tested it in GKE, so maybe there's some more info needed to reproduce this issue @cristifalcas

**Comment by cristifalcas** [comment]
I could put the links to the various issues regarding this. In aws, at least, is happening all the time. Disks are created in multiple azs

**Comment by jeff-knurek** [comment]
ok, so it's a bug in AWS, and the workaround is to switch from deployment to statefulset?

**Comment by cristifalcas** [comment]
I was speaking about this: https://github.com/kubernetes/kubernetes/issues/59326

And here is where they say that it works only for statefulsets:https://github.com/kubernetes/kubernetes/issues/59326#issuecomment-362974193

File: stable/sonatype-nexus/templates/statefulset.yaml [hunk 1]
```diff
@@ -108,6 +108,13 @@ spec:
           {{- end }}
         {{- if .Values.nexusBackup.enabled }}
         - name: nexus-backup
+          annotations:
+            {{- range $key, $value := .Values.ingress.annotations }}
+            {{ $key }}: {{ $value | quote }}
+            {{- end }}
+      {{- if .Values.podAnnotations }}
```
Comment (jeff-knurek) [inline stable/sonatype-nexus/templates/statefulset.yaml:115]: it looks like you mean `.Values.nexusBackup.podAnnotations` ?? 
and I'm not clear on why you're adding the `ingress.annotations` here either

File: stable/sonatype-nexus/templates/statefulset.yaml [hunk 2]
```diff
@@ -146,17 +153,34 @@ spec:
...
+        {{- if not .Values.nexusBackup.persistence.enabled }}
         - name: nexus-backup
-          {{- if and .Values.nexusBackup.enabled (.Values.nexusBackup.persistence.enabled) }}
-          persistentVolumeClaim:
-            claimName: {{ .Values.nexusBackup.persistence.existingClaim | default (printf "%s-%s" (include "nexus.fullname" .) "backup") }}
-          {{- else }}
           emptyDir: {}
           {{- end }}
+{{- if or .Values.persistence.enabled .Values.nexusBackup.persistence.enabled }}
+  volumeClaimTemplates:
```
Comment (jeff-knurek) [inline stable/sonatype-nexus/templates/statefulset.yaml:165]: so by using `volumeClaimTemplates` we're loosing the ability to reuse an existing disk/claim

----

maybe the solution to your issue with AWS and multi-zone disks is to instead first create the disks/claims before applying the helm chart

File: stable/sonatype-nexus/values.yaml [hunk 3]
```diff
@@ -75,6 +75,7 @@ nexusBackup:
   imageName: quay.io/travelaudience/docker-nexus-backup
   imageTag: 1.2.0
   imagePullPolicy: IfNotPresent
+  podAnnotations: {}
```
Comment (jeff-knurek) [inline stable/sonatype-nexus/values.yaml:78]: if you're adding a new configuration option, update the README too please

**Comment by cristifalcas** [comment]
The pod annotation was for extra options for other stuff, for example kube2iam. But since all extra options for this chart are kube cloud centric, we decided to roll out our own chart. So this can be closed now

**Pull request closed without merging**